### PR TITLE
Add the WCPay customer ID to the WP user during checkout

### LINKED ISSUE #2994: Subscriptions with a trial period (or initial payment of zero) fail to renew
### Describe the bug
If the **Enable payments via saved cards** setting is set to off, a subscription with a parent order value of $0 (ie a subscription with a trial period) will save the "payment method" value correctly and will fail to save the "saved payment method" details, even though the parent order will save the `seti` token value correctly.

### To Reproduce

With WooCommerce, WooCommerce Subscriptions, and WooCommerce Payments

1. Set up a subscription product with a recurring total and a free period
2. Set WooCommerce Payments **Enable payments via saved cards** to off
2. Go through the checkout process with initial payment of zero
3. Trigger or wait for the first renewal payment

### Actual behavior

When the parent order and subscriptioon is created

1. The parent order will save the `seti_` token.
2. The Subscription will display WooCommerce Payments as the payment method
3. The Subscription will display no values in the "saved payment method" dropdown
4. The renewal order will fail

### Screenshots

Markup on 2021-09-25 at 07:15:45

Markup 2021-09-25 at 07 13 35

### Expected behavior

I would expect t**Enable payments via saved cards** not to affect subscriptions

### Additional context

reported in 4318150-zen

### PR DESCRIPTION
**Pull request by aprea**
Fixes #2994 

#### Changes proposed in this Pull Request

Hooks into `woocommerce_created_customer` to add the WooComerce Payments customer ID found in the user session to the WordPress user as metadata.

This is helpful in scenarios where the shopper begins the checkout flow logged out (i.e. guest checkout) and a user account is created for them during checkout.

This occurs when a WordPress user account is necessary for checkout, e.g. when the shopper purchases a subscription product.

#### Testing instructions

**Testing setup**
* WooCommerce Subscriptions is **not** active.
* US store with WCPay Subscriptions enabled.
* Create a $40 p/month subscription product that has a 1-week trial period.

**To reproduce the bug**
1. Be on the `develop` branch.
1. Navigate to your store while logged out (you can use an incognito window if it helps).
1. Add the subscription product created above to your cart.
1. Observe the cart total should be $0 (because of the trial) and the recurring total should be $40 p/month.
1. Navigate to checkout.
1. Navigate to your Stripe Connect account dashboard then to Customers.
1. Observe that an empty customer object was created when you landed on the checkout page, the customer description is `Name: , Guest`.
1. Switch back to the checkout page, fill in the required details and submit checkout.
1. Observe the checkout appears to be successful and you are redirected to the order confirmation page.
1. **Bug:** Switch back to the Stripe dashboard, observe a new customer object was created and contains the billing/shipping details you added during checkout but the customer **does not** have a payment method attached.
1. Navigate to the Subscription page in the Stripe dashboard, observe the newly created subscription.
1. **Bug:** Open the subscription, observe that it has the most recent customer object attached but **does not** a payment method attached.

**To observe the fix**
1. Switch to this branch `fix/2994-subscription-trial-fixes`.
2. Follow steps 2-9 from above.
3. Switch back to the Stripe dashboard, observe that the existing `Name: , Guest` customer object was updated with the billing/shipping details you added during checkout and has a payment method attached.
4. Navigate to the Subscription page in the Stripe dashboard, observe the newly created subscription
5. Open the subscription, observe that it has the customer object attached and that it has a payment method attached.

#### TODOs
- Changelog / readme entries.

#### Questions
See the attached screenshot. There are a total of 4 SetupIntents created during this flow. At first glance, this seems a little excessive and I wonder if all of these are necessary? I have also asked here → p1642482576024000/1642054337.047300-slack-C9976E5MJ

**Edit:** I received a response, see below.

>Thanks, based on the flow, really only one setupIntent should be created.
>
>A setupIntent need to be created, in order to render the payment element to the customer to collect their card info.
>
>If a user happen to abandon checkout and then come back later, or refresh page without caching the setupIntent, then more than one setupIntent can be expected

Screen Shot 2022-01-18 at 3 33 44 pm

-------------------

- [ ] Added changelog entries to both `readme.txt` and `changelog.txt` (or does not apply)
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) : _Add link here / 'QA Testing Not Applicable'_

### CODE REVIEW

File: includes/payment-methods/class-upe-payment-gateway.php [hunk 1]
```diff
@@ -380,7 +380,6 @@ public function process_payment( $order_id ) {
 		$currency                  = $order->get_currency();
 		$converted_amount          = WC_Payments_Utils::prepare_amount( $amount, $currency );
 		$payment_needed            = 0 < $converted_amount;
-		$token                     = Payment_Information::get_token_from_request( $_POST ); // phpcs:ignore WordPress.Security.NonceVerification.Missing
```
Comment (aprea) [inline includes/payment-methods/class-upe-payment-gateway.php:383]: Unrelated tidying ✨ This variable was not used in this method.

**Comment by james-allan** [comment]
I've just run through the testing instructions and can replicate the issue and confirm the fix. I want to do some most testing of standard checks but I was curious about the number of intent objects being created. How do you find that list of events?

**Comment by aprea** [comment]
>I was curious about the number of intent objects being created. How do you find that list of events?

You can find the event log here → https://dashboard.stripe.com/[your_account]/test/events

**Comment by james-allan** [comment]
I've tried to find where these additional intent objects are coming from and couldn't find anything that definitive. 1 is linked to the subscription being created - when you create a subscription a pending intent object gets created.

Screen Shot 2022-01-19 at 4 48 59 pm

The image on the right is the subscription create response. 

One of the intent objects (blue above) comes from the original page load - presumably for the UPE element. The other two are a mystery. I logged all server requests and responses and couldn't see where these intent objects come from. 

The orange intent I can't track where it's coming from but I can see that it gets the cancellation reason of "abandoned" 

Screen Shot 2022-01-19 at 4 59 16 pm

The purple one appears to be linked to the payment method being saved to the subscription, but the only reason I think that is because clicking through in the Stripe events shows that it's related to this API request

Screen Shot 2022-01-19 at 5 03 05 pm

When I compare free trial subscriptions with a standard upfront subscription, I notice that 4 payment intent objects (`pi_` vs `seti_`) also get created in this case the multiple intent objects come from:

1. The UPE element generation
2. Initial Stripe billing subscription invoice (the one we just ignore and mark as paid out of band).
3. The actual payment that WC takes to collect the initial payment. 
4. A set up intent linked to saving the payment method to the subscription. 


---- 

I honestly don't know where the 2 extra ones come from when purchasing trial products. I think they would be harmless and might even be coming from Stripe's own processes, but who knows.

**Review by james-allan — APPROVED** [review]
These changes look good to me. I've replicated the initial bug and confirmed this fixes it. 

I haven't been able to discover where the additional setup intent requests are coming from however. 

I tested some other scenarios just to make sure this doesn't introduce any regression. All of those were clear

- Logged in free trial 👍 
- Logged in standard subscription 👍 
- Logged in both subscription trial and upfront in the 1 checkout 👍 
- Logged in standard-simple product 👍 
- Logged out free trial 👍 
- Logged out standard subscription 👍 
- Logged out both subscription trial and upfront in the 1 checkout 👍 
- Logged out standard-simple product 👍

**Comment by aprea** [comment]
Thanks for the review @james-allan 🎉! I might create an investigation issue to look into the excessive SetupIntents.

**Comment by htdat** [comment]
FYI. This PR has been available in WCPay 3.6.0 without going through manual tests done by GlobalStep this week (Jan 18).
p1642611618163400-slack-C01BPL3ALGP

**Pull request merged**